artifact.go

```diff
@@ -267,24 +268,37 @@ func (m artifactModule) postTestResultDataHandler(c *gin.Context) {
 	for _, artifact := range artifacts {
 		summary, details, err := getTestSummaryAndDetails(artifact.Data, artifact.ArtifactID, buildID)
 		if err != nil {
-			log.Warn().
+			logEvent := log.Warn().
 				WithError(err).
 				WithString("filename", artifact.Name).
 				WithUint("build", buildID).
-				WithUint("artifact", artifact.ArtifactID).
-				Message("Failed to unmarshal; invalid TRX/XML format, or the structure" +
-					" receiving the data is malformed.")
-
-			ginutil.WriteProblemError(c, err,
-				problem.Response{
-					Type:   "/prob/unexpected-response-format",
-					Status: 502,
-					Title:  "Unexpected response format.",
-					Detail: fmt.Sprintf(
-						"Failed parsing test result ID %d, for build with ID %d in"+
-							" database. Expected TRX/XML format. Could also be due to an"+
-							" internal bug.", summary.ArtifactID, buildID),
-				})
+				WithUint("artifact", artifact.ArtifactID)
+			if syntaxErr := err.(*xml.SyntaxError); syntaxErr != nil || strings.HasPrefix(err.Error(), "xml:") {
+				logEvent.Message("Failed to unmarshal; invalid/unsupported TRX/XML format.")
+
+				ginutil.WriteProblemError(c, syntaxErr,
+					problem.Response{
+						Type:   "/prob/unexpected-response-format",
+						Status: http.StatusBadGateway,
+						Title:  "Unexpected response format.",
+						Detail: fmt.Sprintf(
+							"Failed parsing test result ID %d, for build with ID %d in"+
+								" database. Invalid/unsupported TRX/XML format.", summary.ArtifactID, buildID),
+					})
+			} else {
+				logEvent.Message("Failed to unmarshal; the structure used to unmarshall might be malformed.")
+
+				ginutil.WriteProblemError(c, syntaxErr,
+					problem.Response{
+						Type:   "/prob/bad-code",
+						Status: http.StatusInternalServerError,
+						Title:  "Bad code.",
+						Detail: fmt.Sprintf(
+							"Failed parsing test result ID %d, for build with ID %d in database. The structure"+
+								" used to unmarshall the data in the wharf-api might be malformed. This really shouldn't happen.",
+							summary.ArtifactID, buildID),
+					})
+			}
```

Comment:
I checked the source for `xml.Unmarshal` to see what errors were returned, and came up with this.
It feels a bit massive, and not sure about the choice for the second problem.

Thoughts?

Reply:
Saw https://github.com/iver-wharf/wharf-api/pull/43#discussion_r695437525 just now.

With the split of plausible error vs. implausible error (in production), I feel that the `Bad code.` and `This really shouldn't happen.` messages are warranted, and makes it clear to both the users and developers that it is something that requires checking the code.

It's likely all errors that would go to the second branch would be found with unit test for that function, allowing us to "ignore" handling those as they wouldn't happen.

Reply:
I'd suggest to change to only check `err != nil` and then return a problem for "Failed to parse TRX test results file" and then let the error itself explain the nitty-gritty reason for it.

You can still have the problem detail be something like it is now with "the data might be malformed" or "Invalid TRX/XML file format".

The problem types here are kind of a stretch IMO. Perhaps instead have `Type: "/prob/api/test-results-parse"` or something, where the docs could suggest that you uploaded a file with the wrong format or that you selected the wrong format in the query parameters or something.